File: webui/src/main/java/com/taskadapter/web/uiapi/UISyncConfigBuilder.java
package com.taskadapter.web.uiapi;

import com.google.gson.Gson;
import com.taskadapter.config.StoredConnectorConfig;
import com.taskadapter.config.StoredExportConfig;
import com.taskadapter.connector.definition.NewMappings;

//TODO: This class should really be package-visible.
// Need to move test to an appropriate package and use it!
public final class UISyncConfigBuilder {

    private final UIConfigService uiConfigService;

    public UISyncConfigBuilder(UIConfigService uiConfigService) {
        this.uiConfigService = uiConfigService;
    }

    /**
     * Create a new UI config instance for a stored config.
     *
     * @param ownerName name of config owner.
     * @param storedConfig stored config to create an instance for.
     * @return new parsed config.
     */
    public UISyncConfig uize(String ownerName, StoredExportConfig storedConfig) {
        final String label = storedConfig.getName();
        final StoredConnectorConfig conn1Config = storedConfig.getConnector1();
        final StoredConnectorConfig conn2Config = storedConfig.getConnector2();
        final UIConnectorConfig config1 = uiConfigService.createRichConfig(conn1Config.getConnectorTypeId(), conn1Config.getSerializedConfig());
        final UIConnectorConfig config2 = uiConfigService.createRichConfig(conn2Config.getConnectorTypeId(), conn2Config.getSerializedConfig());
        final NewMappings mappings = storedConfig.getMappings() == null ? MappingGuesser.guessNewMappings(storedConfig)
                : new Gson().fromJson(storedConfig.getMappings(),
                NewMappings.class);

        // fixing the mappings is important to load config files created by Task Adapter v. 2.2 or older.
        // feel free to delete MappingFixer.fixMappings() call if it's 2013.
//        final NewMappings fixedMappings = MappingFixer.fixMappings(mappings,
//                config1.getAvailableFields(), config2.getAvailableFields(),
//                false);
//        return new UISyncConfig(storedConfig.getId(), ownerName, label,
//                config1, config2, fixedMappings, false);
        return new UISyncConfig(storedConfig.getId(), ownerName, label,
                config1, config2, mappings, false);
    }

}

File: webui/src/main/scala/com/taskadapter/web/uiapi/UISyncConfigBuilder.scala
package com.taskadapter.web.uiapi

import com.taskadapter.config.CirceBoilerplateForConfigs._
import com.taskadapter.config.StoredExportConfig
import com.taskadapter.connector.definition.FieldMapping
import io.circe.parser._

import scala.collection.JavaConverters._

class UISyncConfigBuilder(val uiConfigService: UIConfigService) {
  /**
    * Create a new UI config instance for a stored config.
    *
    * @param ownerName    name of config owner.
    * @param storedConfig stored config to create an instance for.
    * @return new parsed config.
    */
  def uize(ownerName: String, storedConfig: StoredExportConfig): UISyncConfig = {
    val label = storedConfig.getName
    val conn1Config = storedConfig.getConnector1
    val conn2Config = storedConfig.getConnector2
    val config1 = uiConfigService.createRichConfig(conn1Config.getConnectorTypeId, conn1Config.getSerializedConfig)
    val config2 = uiConfigService.createRichConfig(conn2Config.getConnectorTypeId, conn2Config.getSerializedConfig)
    val jsonString = storedConfig.getMappingsString

    val newMappings = decode[Seq[FieldMapping]](jsonString)
    newMappings match {
      case Left(e) => throw new RuntimeException(s"cannot parse mappings from config: $e")
      case Right(m) =>
        new UISyncConfig(storedConfig.getId, ownerName, label, config1, config2, m.asJava, false)
    }
  }
}
